### PR TITLE
[TF-TRT] Remove dependency on einsum kernel

### PR DESCRIPTION
**Pull request by trevor-m**
Previously, TF-TRT op converter for einsum used a helper function from the actual einsum kernel implementation in `tensorflow/core/kernels/linalg/einsum_op_impl.h` and had a dependency on einsum_op. This PR moves the helper function (`ParseEquation`) to `tensorflow/core/util/einsum_op_util.h` and remove such a dependency.

This change makes both the einsum kernel and TF-TRT more modular, which is helpful if we want to move either of them to a separate binary (dynamic kernel library for example).

Summary of changes:
* Rename `ParseEinsumEquation()` to `ValidateEinsumEquation()` in  `tensorflow/core/utils/einsum_op_util.h`
* Rename `EinsumHelper::ParseEquation()` to  `ParseEinsumEquation()` and move from `tensorflow/core/kernels/linalg/einsum_kernel_impl.h` to `tensorflow/core/utils/einsum_op_util.h`.
* Update TF-TRT to use new `ParseEinsumEquation()` in `tensorflow/core/utils/einsum_op_util.h` and remove dependency on `tensorflow/core/kernels/linalg/einsum_kernel_impl.h`.
* Rename `EinsumHelper::DimensionType` to `EinsumDimensionType`.

### CODE REVIEW

**Comment by trevor-m** [comment]
@DEKHTIARJonathan

**Comment by DEKHTIARJonathan** [comment]
@bixia1 for review

File: tensorflow/core/util/einsum_op_util.cc [hunk 1]
```diff
@@ -44,4 +45,100 @@ Status ParseEinsumEquation(const string& equation,
...
+    }
+    const int mapped_label = (*label_mapping)[label_char];
+    labels->push_back(mapped_label);
+  }
+}
+
+// Parses and validates the equation and the input shapes. Single character
+// labels are integerized and we populate input and output label subscripts
+// and corresponding counts. Also create the mapping from (named) labels to
+// their EinsumDimensionType.
```
Comment (bixia1) [inline tensorflow/core/util/einsum_op_util.cc:85]: Let's move this comment to the header file.

File: tensorflow/core/util/einsum_op_util.h [hunk 2]
```diff
@@ -20,10 +20,49 @@ limitations under the License.
...
+
+// Each dimension is categorized into exactly one of five types based on
+// whether its corresponding label is present in the input and/or the output
+// subscripts.
+enum EinsumDimensionType {
+  // Batch dimensions are those present in two inputs as well as the output.
+  // They are part of the batch dimensions during Tensor contraction.
+  // Such dimensions may be broadcasting dimensions (those mapping to
+  // ellipsis)
+  // or explicit batch dimensions corresponding to named axis labels.
```
Comment (bixia1) [inline tensorflow/core/util/einsum_op_util.h:39]: Please re-organize these comment lines to fill all lines except for the last line.

File: tensorflow/core/util/einsum_op_util.cc [hunk 3]
```diff
@@ -44,4 +45,100 @@ Status ParseEinsumEquation(const string& equation,
...
+                           OperandLabelCounts* input_label_counts,
+                           LabelCounts* output_label_counts,
+                           gtl::InlinedVector<bool, 2>* input_has_ellipsis,
+                           bool* output_has_ellipsis) {
+  gtl::InlinedVector<string, 2> input_str;
+  string output_str;
+  TF_RETURN_IF_ERROR(ValidateEinsumEquation(equation, &input_str, &output_str));
+
+  // Temporary map from single character labels to (consecutive) integer
+  // labels.
```
Comment (bixia1) [inline tensorflow/core/util/einsum_op_util.cc:98]: "labels." can go into the previous line?

**Comment by bixia1** [comment]
@trevor-m I modified your PR description, please check.

**Comment by trevor-m** [comment]
Thank you @bixia1 for the review! I have addressed your comments.

**Review by bixia1 — APPROVED** [review]

**Pull request merged**